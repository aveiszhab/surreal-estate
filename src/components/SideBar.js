import React, { useState } from "react";
import "../styles/SideBar.css";
import { Link, useLocation } from "react-router-dom";
import qs from "qs";

const SideBar = () => {
  const { search } = useLocation();
  const buildQueryString = (operation, valueObj) => {
    const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });

    const newQueryParams = {
      ...currentQueryParams,
      [operation]: JSON.stringify(valueObj),
    };

    return qs.stringify(newQueryParams, {
      addQueryPrefix: true,
      encode: false,
    });
  };
  return (
    <div>
      <h2>Filter by City</h2>
      <div className="sidebar-links">
        <div className="sidebar-links-items">
          <Link to={buildQueryString("query", { city: "Manchester" })}>
            Manchester
          </Link>
        </div>
        <div className="sidebar-links-items">
          <Link to={buildQueryString("query", { city: "Leeds" })}>Leeds</Link>
        </div>
        <div className="sidebar-links-items">
          <Link to={buildQueryString("query", { city: "Sheffield" })}>
            Sheffield
          </Link>
        </div>
        <div className="sidebar-links-items">
          <Link to={buildQueryString("query", { city: "Liverpool" })}>
            Liverpool
          </Link>
        </div>
      </div>
      <h2>Sort by Price</h2>
      <div className="sidebar-links">
        <div className="sidebar-links-items">
          <Link to={buildQueryString("sort", { price: +1 })}>
            Price Ascending
          </Link>
        </div>
        <div className="sidebar-links-items">
          <Link to={buildQueryString("sort", { price: -1 })}>
            Price Descending
          </Link>
        </div>
      </div>
    </div>
  );
};

export default SideBar;
