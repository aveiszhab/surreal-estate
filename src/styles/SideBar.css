.sidebar-links {
    display: flex;
    flex-direction: column;
    text-align: left;
}

.sidebar-links-items {
    margin-left: 5%;
    padding: 5px 0%;
    text-decoration: none;
    color: ivory;
}

h4 {
    margin-left: 5%;
}

.search {
    display: flex;
    margin: 5%;
    justify-content: center;
}

.search-input {
   width: 65%;
   padding: 6%;
   border-radius: 8px 0px 0px 8px;
   background-color:ivory;
   border: 1px solid rgb(9, 32, 2);
   font-weight: 600;

}

.search-button {
    background-color:ivory;
    border-radius: 0px 8px 8px 0px;
    border: 1px solid rgb(9, 32, 2);
    cursor: pointer;
    padding: 0% 5%;
    font-weight: 400;
    justify-content: center;
}





