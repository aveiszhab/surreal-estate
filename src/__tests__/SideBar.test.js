import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from "react-router-dom";
import SideBar from "../components/SideBar";

describe("SideBar", () => {
  const setUp = () => {
    const { asFragment, getByRole } = render(
      <Router>
        <SideBar />
      </Router>
    );

    return {
      asFragment: asFragment(),
      filterHeading: getByRole("heading", { name: /filter/i }),
      sortHeading: getByRole("heading", { name: /sort/i }),
      manchester: getByRole("link", { name: /manchester/i }),
      leeds: getByRole("link", { name: /leeds/i }),
      sheffield: getByRole("link", { name: /sheffield/i }),
      liverpool: getByRole("link", { name: /liverpool/i }),
      ascending: getByRole("link", { name: /ascending/i }),
      descending: getByRole("link", { name: /descending/i }),
      searchButton: getByRole("button", { name: /search/i }),
      resetButton: getByRole("button", { name: /clear/i }),
      searchInput: getByRole("textbox", { name: /search-input/i }),
    };
  };

  let renderedComponent;

  beforeEach(() => {
    renderedComponent = setUp();
  });

  it("renders correctly to match the SideBar snapshot", () => {
    expect(renderedComponent.asFragment).toMatchSnapshot();
  });

  it("renders correctly", () => {
    expect(renderedComponent.filterHeading).toHaveTextContent("Filter by City");
    expect(renderedComponent.sortHeading).toHaveTextContent("Sort by Price");

    expect(renderedComponent.manchester).toHaveTextContent("Manchester");
    expect(renderedComponent.manchester).toHaveAttribute(
      "href",
      '/?query={"city":"Manchester"}'
    );
    expect(renderedComponent.sheffield).toHaveTextContent("Sheffield");
    expect(renderedComponent.sheffield).toHaveAttribute(
      "href",
      '/?query={"city":"Sheffield"}'
    );

    expect(renderedComponent.leeds).toHaveTextContent("Leeds");
    expect(renderedComponent.leeds).toHaveAttribute(
      "href",
      '/?query={"city":"Leeds"}'
    );

    expect(renderedComponent.liverpool).toHaveTextContent("Liverpool");
    expect(renderedComponent.liverpool).toHaveAttribute(
      "href",
      '/?query={"city":"Liverpool"}'
    );

    expect(renderedComponent.ascending).toHaveTextContent(/ascending/i);
    expect(renderedComponent.ascending).toHaveAttribute(
      "href",
      '/?sort={"price":1}'
    );

    expect(renderedComponent.descending).toHaveTextContent(/descending/i);
    expect(renderedComponent.descending).toHaveAttribute(
      "href",
      '/?sort={"price":-1}'
    );

    expect(renderedComponent.searchButton).toHaveClass("search-button");
    expect(renderedComponent.searchButton).toBeEnabled();

    expect(renderedComponent.resetButton).toHaveClass("reset-button");
    expect(renderedComponent.resetButton).toBeEnabled();

    expect(renderedComponent.searchInput).toHaveClass("search-input");
  });
});
