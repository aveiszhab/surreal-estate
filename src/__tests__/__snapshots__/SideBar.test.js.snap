// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar renders correctly to match the SideBar snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      Search by Title
    </h4>
    <form
      class="search"
    >
      <input
        class="search-input"
        title="search-input"
        type="text"
        value=""
      />
      <button
        class="search-button"
        title="search"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </form>
    <div
      class="sidebar-links"
    >
      <h4>
        Filter by City
      </h4>
      <a
        class="sidebar-links-items"
        href="/?query={\\"city\\":\\"Manchester\\"}"
      >
        Manchester
      </a>
      <a
        class="sidebar-links-items"
        href="/?query={\\"city\\":\\"Leeds\\"}"
      >
        Leeds
      </a>
      <a
        class="sidebar-links-items"
        href="/?query={\\"city\\":\\"Sheffield\\"}"
      >
        Sheffield
      </a>
      <a
        class="sidebar-links-items"
        href="/?query={\\"city\\":\\"Liverpool\\"}"
      >
        Liverpool
      </a>
      <h4>
        Sort by Price
      </h4>
      <a
        class="sidebar-links-items"
        href="/?sort={\\"price\\":1}"
      >
        Price Ascending
      </a>
      <a
        class="sidebar-links-items"
        href="/?sort={\\"price\\":-1}"
      >
        Price Descending
      </a>
    </div>
    <button
      class="reset-button"
      type="submit"
    >
      Clear Filters
    </button>
  </div>
</DocumentFragment>
`;
